File: Heapsort/heapsort.java
import java.util.*;

class heapsort {
	public static void main(String args[]) {
		ArrayList<Integer> myArr = new ArrayList<Integer>();
		myArr.add(0);
		myArr.add(5);
		myArr.add(4);
		myArr.add(3);
		heapsort(myArr);
		print(myArr);
	}

	private static void print(List<?> list) {
		for (int i = 1; i < list.size(); i++) {
			System.out.print(list.get(i) + " ");
		}
		System.out.println();
	}

	private static <T extends Comparable> void heapsort(List<T> list) {
		buildHeap(list);
		int heapSize = list.size();
		for (int i = 1; i < list.size(); i++) {
			T last = list.get(heapSize - 1);
			list.set(heapSize - 1, list.get(1));
			list.set(1, last);
			percolateDown(list, 1, --heapSize);
		}
	}

	private static <T extends Comparable> void buildHeap(List<T> list) {
		for (int i = (list.size() - 1) / 2; i > 0; i--) {
			percolateDown(list, i, list.size());
		}
	}

	private static <T extends Comparable> void percolateDown(List<T> list, int hole, int currentSize) {
		T temp = list.get(hole);
		int child;

		for (; hole * 2 <= currentSize - 1; hole = child) {
			child = hole * 2;
			if (child != currentSize - 1 && list.get(child + 1).compareTo(list.get(child)) > 0) {
				++child;
			}
			if (list.get(child).compareTo(temp) > 0) {
				list.set(hole, list.get(child));
			}
			else {
				break;
			}
		}
		list.set(hole, temp);
	}
}

File: Heapsort/heapsort.go
package main
import "fmt"

func percolateDown(list []int, hole int, currentSize int) {
	var child int
	var temp int = list[hole]

	for ; hole * 2 <= currentSize - 1; hole = child {
		child = hole * 2
		if (child != currentSize - 1 && list[child + 1] > list[child]) {
			child++
		}

		if (list[child] > temp) {
			list[hole] = list[child]
		} else {
			break
		}
	}
	list[hole] = temp
}

func buildHeap(list []int) {
	for i := (len(list) - 1) / 2; i > 0; i-- {
		percolateDown(list, i, len(list))
	}
}

func heapsort(list []int) {
	buildHeap(list)
	var heapSize int = len(list)
	for i := 1; i < len(list); i++ {
		last := list[heapSize - 1]
		list[heapSize - 1] = list[1]
		list[1] = last
		heapSize--
		percolateDown(list, 1, heapSize)
	}
}

func print(list []int) {
	for i := 1; i < len(list); i++ {
		fmt.Print(list[i], " ")
	}
	fmt.Println()
}

func main() {
	list := []int{0, 1, 2, 3, 67, 32, 2, 5, 8, 4, 5, 6, 7}
	heapsort(list)
	print(list)
}

